### DISCUSSION TITLE
csr과 ssr에 대해서 설명해주세요.

### DISCUSSION THREAD

**Discussion by hwangstar156** [woowacourse-study/2022-woowahan-bansanghwe#100] [JavaScript]
## CSR

CSR(Client Side Rendering)은 데이터가 없는 빈 깡통인 HTML(이외에 static 파일들)만 받아오고, 데이터는 깡통인 문서와 여러 스테틱 파일들이 로드된 이후에 요청해서 받아오는 방식으로 진행됩니다.

즉, 사용자의 요청에 따라 필요한 부분만 응답 받아 렌더링 하는 방식이라고 할 수 있습니다.

### 과정

먼저 클라이언트에서 초기화면을 로드하기 위해 서버에 요청을 보냅니다.

그럼 서버 는 화면에 표시하는 데 필요한 완전한 리소스를 응답합니다.

그런데 여기서 클라이언트 사이드 렌더링 방식이 서버사이드 렌더링 방식과 다른 점은 모든 js 파일을 다운받아 하기 때문에 초기 로딩 시간의 더 오래 걸린다는 점입니다.

### 장점

- 페이지 이동시 요청을 하지 않기에 빠른 속도의 응답을 받을수 있음.
- 서버부하가 줄어든다.

### 단점
- SEO가 불리하다.
- 느린 초기로딩

### 과정

ssr이란 서버사이드 렌더링의 약자로 서버로부터 완전하게 만들어진 html파일을 받아와 페이지 전체를 렌더링 하는 방식입니다.

먼저 클라이언트가 초기 화면을 로드하기위해 서버에 요청을 보냅니다.

그럼 서버는 화면에 표시하는데 필요한 데이터를 얻어와 모두 삽입하고 css 까지 모두 적용해서 렌더링 준비를 마친 HTML과 JS코드를 브라우저에 응답으로 전달합니다.

브라우저에서는 바로 전달 받은 페이지를 띄우고 이어, 브라우저가 JS 코드를 다운로드하고 html에 실행시킵니다.

### 장점
- 검색엔진 최적화
- 빠른 초기 로딩

### 단점
- 페이지 이동시 마다 요청을 보내 서버부하가 생김
- 페이지 이동시 마다 요청을 보내기 떄문에 새로고침(깜빡임) 현상 발생

**Comment by JUDONGHYEOK** [comment]
CSR과 SSR에 대해서 설명해주세요

**Comment by JUDONGHYEOK** [comment]
CSR과 SSR을 사용해본 경험

**Comment by JUDONGHYEOK** [comment]
그렇게 했을 때 장점

**Comment by JUDONGHYEOK** [comment]
React hydrate를 알고있나?

**Comment by JUDONGHYEOK** [comment]
react는 왜 csr을 채택했나?

**Comment by JUDONGHYEOK** [comment]
SSG, ISR를 알고있나?

**Comment by JUDONGHYEOK** [comment]
브라우저 렌더링과정 설명

**Comment by JUDONGHYEOK** [comment]
리플로우, 리페인트, 컴포지트 단계를 세세하게 이야기해줄 수 있나?

**Comment by JUDONGHYEOK** [comment]
컴포지트과정에서는 어떤 일이 일어나나?

**Comment by JUDONGHYEOK** [comment]
리액트의 생명주기 설명해주세요

**Comment by JUDONGHYEOK** [comment]
리액트는 왜 클래스형 컴포넌트를 채택했을까?

**Comment by JUDONGHYEOK** [comment]
왜 리액트를 사용할까?

**Comment by JUDONGHYEOK** [comment]
리액트를 사용하면서 단점
